### DISCUSSION TITLE
Issue with persist example in docs?

### DISCUSSION THREAD

**Discussion by ConnorLanglois** [pmndrs/zustand#2251] [Bug report] [answered]
Hi I noticed that in the faq section of the persist docs for checking if the store is hydrated ([located here](https://docs.pmnd.rs/zustand/integrations/persisting-store-data#how-can-i-check-if-my-store-has-been-hydrated)) there seems to be an issue. The code given is
```js
const useBoundStore = create(
  persist(
    (set, get) => ({
      // ...
      _hasHydrated: false,
      setHasHydrated: (state) => {
        set({
          _hasHydrated: state
        });
      }
    }),
    {
      // ...
      onRehydrateStorage: () => (state) => {
        state.setHasHydrated(true)
      }
    }
  )
);
```

Won't this persist the "_hasHydrated" field itself as "true" into the storage backend, and then during hydration on next launch it will be true always? Shouldn't partialize be used to prevent this? Or is there a special case with fields that start with an underscore?

Thank you

**Comment by dai-shi** [comment]
> Shouldn't partialize be used to prevent this?

That sounds about right.

Would anyone be able to investigate and confirm it? cc @dbritto-dev @sewera @charkour

**Reply by dbritto-dev** [reply]
I'll look into that

**Reply by charkour** [reply]
@dbritto-dev, I can also look into it if you'd appreciate it.

**Reply by dai-shi** [reply]
Super glad to see active zustand team.

**Comment by charkour** [comment, marked as answer]
Hi @ConnorLanglois,

Thank you for your questions! I use zustand a lot but only have used `persist` some. I'll do my best to answer your questions. 

> Won't this persist the "_hasHydrated" field itself as "true" into the storage backend, and then during hydration on next launch it will be true always?

This is a good question. In this example, the default storage option in `localStorage`, so when this store is created in JS, the default value of `_hasHydrated` will be `true` because it is synchronous storage, but if you were to use asynchronous storage, it would be set to `false` until the persisted storage object is merged with the current state. The current state when the store is created will always be `_hasHydrated === false`. I recommend reviewing these [docs](https://docs.pmnd.rs/zustand/integrations/persisting-store-data#hydration-and-asynchronous-storages). 

> Shouldn't partialize be used to prevent this?

`partialize` is used to prevent specific states from being persisted in storage. While we could remove `_hasHydrated` from being stored via the `partialize` option, it's not necessary because `_hasHydrated` will be `false` if the store hasn't hydrated, and will be set to `true` once hydration has happened (meaning the persisted state has merged with the current state). 

> Or is there a special case with fields that start with an underscore?

I looked through the [source code](https://github.com/pmndrs/zustand/blob/main/src/middleware/persist.ts), and didn't find any special cases for underscores. 

@dbritto-dev, please add on anything I may have missed. Thanks

**Reply by ConnorLanglois** [reply]
Thanks for the reply! I re-read the code and your response and it makes sense now. Was confused that it writes it to storage (which is bit odd since it's just used as a sort of meta in-memory variable, not meant to be actually stored) but I can see that it should work because the storage is async and that variable starts at `false`, makes sense. It's interesting, seems broken and correct at same time in my head haha

**Reply by charkour** [reply]
I agree, persist and hydration are confusing!
